File: tests/test_tutorials.R
# CoastWatch satellite-course R tutorials, run against xpublish-erddap.
# Run by CI after starting tests/server.py; see .github/workflows/tests.yml.
#
# The data-access steps are the tutorials' own, with only the server swapped.
# The dataset is the stand-in from tests/tutorial_data.py: the real
# CRW_sst_v1_0_monthly axes, with formula values (sst() below).
suppressMessages({
  library(httr)
  library(ncdf4)
  library(rerddap)
  library(rerddapXtracto)
})

server <- "http://127.0.0.1:9000/erddap/"
ok <- function(msg) cat("ok -", msg, "\n")

# tests/tutorial_data.py::sst, in R
sst_formula <- function(time, lat, lon) {
  lt <- as.POSIXlt(time, tz = "GMT")
  month <- ((lt$year - 70) * 12 + lt$mon) %% 12
  28 - 20 * (lat / 90)^2 + cos(2 * pi * month / 12) + 0.01 * lon
}

# --- Tutorial 1: how to work with satellite data in R -----------------------
url <- paste0(
  server, "griddap/CRW_sst_v1_0_monthly.nc?analysed_sst",
  "[(2018-01-01T12:00:00Z):1:(2018-12-01T12:00:00Z)][(17):1:(30)][(195):1:(210)]"
)
path <- file.path(tempdir(), "sst.nc")
junk <- GET(url, write_disk(path, overwrite = TRUE))
stopifnot(status_code(junk) == 200)

nc <- nc_open(path)
stopifnot("analysed_sst" %in% names(nc$var))
v1 <- nc$var[[1]]
sst <- ncvar_get(nc, v1)
stopifnot(identical(dim(sst), c(301L, 261L, 12L)))
dates <- as.POSIXlt(v1$dim[[3]]$vals, origin = "1970-01-01", tz = "GMT")
lon <- v1$dim[[1]]$vals
lat <- v1$dim[[2]]$vals
nc_close(nc)
ok("tutorial 1: GET + nc_open give a 301 x 261 x 12 array")

stopifnot(format(dates[1], "%Y-%m-%d %H") == "2018-01-01 12")
stopifnot(format(dates[12], "%Y-%m-%d %H") == "2018-12-01 12")
stopifnot(abs(min(lat) - 17) < 0.05, abs(max(lon) - 210) < 0.05)
ok("tutorial 1: dates, latitudes and longitudes are as requested")

grid <- expand.grid(lon = lon, lat = lat, t = seq_along(dates))
expected <- sst_formula(dates[grid$t], grid$lat, grid$lon)
stopifnot(max(abs(as.vector(sst) - expected)) < 1e-6)
ok("tutorial 1: every value matches")

# --- Tutorial 3: extract data within a shapefile using ERDDAP ---------------
# The tutorial uses goes-poes-monthly-ghrsst-RAN; the stand-in is the CRW set.
dataInfo <- rerddap::info("CRW_sst_v1_0_monthly", url = server)
parameter <- dataInfo$variable$variable_name[1]
stopifnot(parameter == "analysed_sst")
ok("tutorial 3: info() names the parameter")

# A polygon across the dateline, in 0-360 longitudes as the tutorial converts
# them to, standing in for the Papahanaumokuakea monument shapefile.
xcoord <- c(177.5, 181.0, 190.0, 199.0, 198.0, 183.0, 177.5)
ycoord <- c(28.5, 29.5, 26.0, 22.0, 20.0, 25.0, 28.5)
tcoord <- c("2015-03-15", "2015-11-16")

sst <- rxtractogon(
  dataInfo, parameter = parameter,
  xcoord = xcoord, ycoord = ycoord, tcoord = tcoord
)
values <- sst[[parameter]]
stopifnot(length(dim(values)) == 3, dim(values)[3] == 9)
stopifnot(any(is.na(values)), any(!is.na(values)))
ok(sprintf(
  "tutorial 3: rxtractogon() returned %s, masked outside the polygon",
  paste(dim(values), collapse = " x ")
))

times <- as.POSIXct(sst$time, tz = "GMT")
grid <- expand.grid(lon = sst$longitude, lat = sst$latitude, t = seq_along(times))
expected <- sst_formula(times[grid$t], grid$lat, grid$lon)
inside <- !is.na(as.vector(values))
stopifnot(max(abs(as.vector(values)[inside] - expected[inside])) < 1e-6)
# rxtracto reads the axes itself and pads to the enclosing grid cells
stopifnot(min(sst$longitude) > 177.5 - 0.05, max(sst$longitude) < 199.0 + 0.05)
stopifnot(min(sst$longitude) < 180, max(sst$longitude) > 180)
ok("tutorial 3: values inside the polygon match, across the dateline")

cat("\nAll tutorial checks passed\n")

File: tests/test_tutorials.py
"""User tutorials, run unchanged against xpublish-erddap.

Each test repeats the data-access steps of a tutorial our users follow, with
only the server swapped. The datasets are stand-ins from ``tutorial_data.py``:
the real ids and axes, with formula values, so every value can be checked.

- CoastWatch satellite course, Python tutorials 1 and 3
  (coastwatch.gitbook.io/satellite-course): a hand-built griddap ``.nc`` URL
  downloaded with ``urllib``, opened with ``decode_cf=False``, times decoded
  with ``netCDF4.num2date``.
- erddapy docs, ``01a-griddap-output``: ``griddap_initialize``, strides, a
  0-360 bounding box, ``to_xarray()``.

The R tutorials are in ``test_tutorials.R``.
"""

import sys
import urllib.request

import numpy as np
import pytest
import xarray as xr
from tutorial_data import rose, sst

erddapy = pytest.importorskip("erddapy")
nc = pytest.importorskip("netCDF4")

pytestmark = pytest.mark.skipif(
    sys.platform == "win32",
    reason="Live server tests are unreliable on Windows Github Actions workers",
)

ORIGINAL = "https://oceanwatch.pifsc.noaa.gov/erddap"


def _download(url: str, server: str, tmp_path) -> xr.Dataset:
    """The tutorials' download-then-open steps, pointed at our server."""
    path = tmp_path / "sst.nc"
    urllib.request.urlretrieve(url.replace(ORIGINAL, server), path)  # noqa: S310
    return xr.open_dataset(path, decode_cf=False)


def _check_sst(ds: xr.Dataset) -> None:
    """Every value matches the stand-in's formula at the returned coordinates."""
    times = np.array(
        nc.num2date(ds.time.values, ds.time.units, only_use_cftime_datetimes=False),
        dtype="datetime64[ns]",
    )
    expected = sst(
        times[:, None, None],
        ds.latitude.values[None, :, None],
        ds.longitude.values[None, None, :],
    )
    np.testing.assert_allclose(ds.analysed_sst.values, expected)


def test_python_tutorial_1(xpublish_server, tmp_path):
    """CoastWatch Python tutorial 1: how to work with satellite data."""
    url = (
        "https://oceanwatch.pifsc.noaa.gov/erddap/griddap/CRW_sst_v1_0_monthly.nc"
        "?analysed_sst[(2018-01-01T12:00:00Z):1:(2018-12-01T12:00:00Z)]"
        "[(17):1:(30)][(195):1:(210)]"
    )
    ds = _download(url, xpublish_server, tmp_path)

    # the tutorial prints this shape
    assert ds.analysed_sst.shape == (12, 261, 301)

    dates = nc.num2date(ds.time, ds.time.units)
    assert [d.strftime("%Y-%m-%d %H") for d in dates] == [
        f"2018-{m:02d}-01 12" for m in range(1, 13)
    ]

    lat_bnds, lon_bnds = [18, 23], [200, 206]
    da = ds.sel(latitude=slice(*lat_bnds), longitude=slice(*lon_bnds))
    # grid points sit at .025 offsets, so 18..23 holds 100 of them
    assert da.analysed_sst.shape == (12, 100, 120)
    _check_sst(ds)


def test_python_tutorial_3(xpublish_server, tmp_path):
    """CoastWatch Python tutorial 3: extract data within a shapefile."""
    url = (
        "https://oceanwatch.pifsc.noaa.gov/erddap/griddap/CRW_sst_v1_0_monthly.nc"
        "?analysed_sst[(2019-01-15):1:(2019-12-15)]"
        "[(19.2345832):1:(31.79786423)][(177.84422):1:(198.9827)]"
    )
    ds = _download(url, xpublish_server, tmp_path)

    # the tutorial prints this shape
    assert ds.analysed_sst.shape == (12, 252, 424)
    # date-only times snap to the nearest monthly stamp
    dates = nc.num2date(ds.time, ds.time.units)
    assert dates[0].strftime("%Y-%m-%d") == "2019-01-01"
    assert dates[-1].strftime("%Y-%m-%d") == "2019-12-01"
    _check_sst(ds)


@pytest.fixture
def etopo(xpublish_server):
    """The erddapy example's client, pointed at our server."""
    e = erddapy.ERDDAP(server=xpublish_server, protocol="griddap")
    e.dataset_id = "etopo5_EDDGridCopy"
    e.griddap_initialize()
    return e


def test_erddapy_griddap_example_defaults(etopo):
    """The example prints its variables and default constraints.

    These are the strings the erddapy docs show for the real server.
    """
    assert etopo.variables == ["ROSE"]
    assert etopo.constraints == {
        "latitude>=": "-90.0",
        "latitude<=": "90.0",
        "latitude_step": "1",
        "longitude>=": "0.0",
        "longitude<=": "359.91999999999996",
        "longitude_step": "1",
    }


def test_erddapy_griddap_example_download(etopo):
    """Strides of 10, then a 0-360 bounding box, then to_xarray()."""
    etopo.constraints["latitude_step"] = 10
    etopo.constraints["longitude_step"] = 10
    ds = etopo.to_xarray()
    assert ds.ROSE.shape == (217, 432)

    etopo.constraints.update(
        {
            "longitude>=": 290.908,
            "longitude<=": 340.365,
            "latitude>=": -60.533,
            "latitude<=": 0.033,
        },
    )
    ds = etopo.to_xarray()
    assert float(ds.latitude.min()) == pytest.approx(-60.5, abs=0.05)
    assert float(ds.longitude.min()) == pytest.approx(290.9, abs=0.05)
    assert ds.latitude.max() <= 0.033
    assert ds.longitude.max() <= 340.365
    np.testing.assert_allclose(
        ds.ROSE.values,
        rose(ds.latitude.values[:, None], ds.longitude.values[None, :]),
    )


@pytest.mark.xfail(
    reason="response='opendap' needs .dods, which is not served yet (#2)",
    strict=True,
)
def test_erddapy_opendap_response(etopo):
    """erddapy can also open the griddap URL itself as OPeNDAP."""
    etopo.response = "opendap"
    ds = etopo.to_xarray()
    assert ds.ROSE.shape == (2161, 4320)
    assert float(ds.ROSE[1080, 0]) == 0.0
